ReactNativeClient/lib/services/searchengine/queryBuilder.ts

```diff
@@ -0,0 +1,361 @@
+const { time } = require('lib/time-utils.js');
+
+const tagFilter = (tags: string[], intersect: boolean = true, negated: boolean = false) => { // note_tags.tag_id  (NOT) IN (SELECT
+	let result = new Array(tags.length).fill(`${intersect ? 'INTERSECT' : 'UNION'} SELECT note_tags.note_id FROM note_tags WHERE note_tags.tag_id IN (SELECT tags.id from tags WHERE tags.title LIKE ?)`).join(' ');
+	result = intersect ? `SELECT note_tags.note_id as id FROM note_tags WHERE 1 ${result}` : `SELECT note_tags.note_id as id FROM note_tags WHERE 0 ${result}`;
+	result = negated ? `tag_filter_negated as (${result})` : `tag_filter as (${result})`;
+	return result;
+};
+
+
+// WITH RECURSIVE resource_filter as (
+// 	SELECT note_resources.note_id as id FROM note_resources WHERE 1 INTERSECT SELECT note_resources.note_id FROM note_resources WHERE note_resources.is_associated=1
+// 	AND note_resources.resource_id IN (SELECT resources.id from resources WHERE resources.mime LIKE 'application/pdf'))
+
+const resourceFilter = (mimeTypes: string[], intersect: boolean = true, negated: boolean = false) => {
+	let result = new Array(mimeTypes.length).fill(`${intersect ? 'INTERSECT' : 'UNION'} SELECT note_resources.note_id FROM note_resources WHERE note_resources.is_associated=1 AND note_resources.resource_id IN (SELECT resources.id from resources WHERE resources.mime LIKE ?)`).join(' ');
+	result = intersect ? `SELECT note_resources.note_id as id FROM note_resources WHERE 1 ${result}` : `SELECT note_resources.note_id as id FROM note_resources WHERE 0 ${result}`;
+	result = negated ? `resource_filter_negated as (${result})` : `resource_filter as (${result})`;
+	return result;
+};
+
+const negatedAnyResourceFilter = (mimeTypes: string[]) => {
+	let result = new Array(mimeTypes.length).fill('select * from (select * from notes_with_resources except SELECT note_resources.note_id FROM note_resources WHERE note_resources.is_associated=1 AND note_resources.resource_id IN (select resources.id from resources where resources.mime like ?))').join(' UNION ');
+	result = `resource_filter_negated(id) as (${result})`;
+	return result;
+};
+
```

Comment:
Maybe it's just me but I'm having a hard time making sense of what's in this file and I feel it will be hard to maintain this over the long term. First step would be to put the SQL into multiple lines.

Another thing you could try is to not use any `let` variables, only `const` and give meaningful names to each part of the query. Then you return a concatenation of all these `const` variables. Doing so means you'll need to restructure the code in a way that makes it easier to understand each part.